File: Xpotify/SpotifyApi/Library.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace Xpotify.SpotifyApi
{
    public class Library : ApiBase
    {
        public async Task<bool> IsTrackSaved(string trackId)
        {
            return (await IsTrackSaved(new[] { trackId }))[0];
        }

        public async Task<bool[]> IsTrackSaved(string[] trackIds)
        {
            List<bool> output = new List<bool>();

            for (int i = 0; i < trackIds.Length; i += 50)
            {
                var slice = trackIds.Skip(i).Take(50).ToArray();
                output.AddRange(await IsTrackSavedInternal(slice));
            }

            return output.ToArray();
        }

        private async Task<bool[]> IsTrackSavedInternal(string[] trackIds)
        {
            var result = await SendRequestWithTokenAsync(
                "https://api.spotify.com/v1/me/tracks/contains?" +
                $"ids={string.Join(',', trackIds)}", HttpMethod.Get);

            var resultString = await result.Content.ReadAsStringAsync();

            return JsonConvert.DeserializeObject<bool[]>(resultString);
        }

        public async Task<bool> SaveTrackToLibrary(string trackId)
        {
            var result = await SendRequestWithTokenAsync(
                "https://api.spotify.com/v1/me/tracks?" +
                $"ids={trackId}", HttpMethod.Put);

            return result.IsSuccessStatusCode;
        }

        public async Task<bool> RemoveTrackFromLibrary(string trackId)
        {
            var result = await SendRequestWithTokenAsync(
                "https://api.spotify.com/v1/me/tracks?" +
                $"ids={trackId}", HttpMethod.Delete);

            return result.IsSuccessStatusCode;
        }
    }
}

File: Xpotify/Scripts/SpotifyApi/library.ts
namespace XpoMusicScript.SpotifyApi {
    export class Library extends ApiBase {
        public async isTrackSaved(trackId: string): Promise<boolean> {
            return (await this.isTracksSaved([trackId]))[0];
        }

        public async isTracksSaved(trackIds: string[]): Promise<boolean[]> {
            var output = [];
            for (var i = 0; i < trackIds.length; i += 50) {
                var slice = trackIds.slice(i, i + 50);
                output = output.concat(await this.isTracksSavedInternal(slice));
            }
            return output;
        }

        private async isTracksSavedInternal(trackIds: string[]): Promise<boolean[]> {
            var url = 'https://api.spotify.com/v1/me/tracks/contains?ids=' + trackIds.join(',');
            var result = await this.sendJsonRequestWithToken(url, 'get');

            var data = await result.json();
            return data;
        }

        public async saveTrack(trackId: string): Promise<boolean> {
            var url = "https://api.spotify.com/v1/me/tracks?ids=" + trackId;
            var result = await this.sendJsonRequestWithToken(url, 'put');

            return result.status >= 200 && result.status <= 299;
        }

        public async removeTrack(trackId: string): Promise<boolean> {
            var url = "https://api.spotify.com/v1/me/tracks?ids=" + trackId;
            var result = await this.sendJsonRequestWithToken(url, 'delete');

            return result.status >= 200 && result.status <= 299;
        }
    }
}